crates/napi/src/bindgen_runtime/js_values/promise.rs

```diff
@@ -0,0 +1,102 @@
+use std::ffi::CString;
+use std::fmt::Debug;
+use std::future;
+use std::pin::Pin;
+use std::ptr;
+use std::task::{Context, Poll};
+
+use tokio::sync::oneshot::{channel, Receiver, Sender};
+
+use crate::{check_status, Error, Result, Status};
+
+use super::FromNapiValue;
+
+pub struct Promise<T: FromNapiValue> {
+  value: Pin<Box<Receiver<*mut Result<T>>>>,
+}
+
+unsafe impl<T: FromNapiValue> Send for Promise<T> {}
+unsafe impl<T: FromNapiValue> Sync for Promise<T> {}
+
+impl<T: FromNapiValue + Debug> FromNapiValue for Promise<T> {
+  unsafe fn from_napi_value(
+    env: napi_sys::napi_env,
+    napi_val: napi_sys::napi_value,
+  ) -> crate::Result<Self> {
+    let mut then = ptr::null_mut();
+    check_status!(
+      napi_sys::napi_get_named_property(env, napi_val, CString::new("then")?.as_ptr(), &mut then,),
+      "Failed to get then function"
+    )?;
+
+    let mut then_js_cb = ptr::null_mut();
+    let (tx, rx) = channel();
+    check_status!(
+      napi_sys::napi_create_function(
+        env,
+        CString::new("then")?.as_ptr(),
```

Comment:
This will cause UB I think?  Looks like the example in [std docs](https://doc.rust-lang.org/std/ffi/struct.CString.html#method.as_ptr)

```rust
use std::ffi::CString;

let ptr = CString::new("Hello").expect("CString::new failed").as_ptr();
unsafe {
    // `ptr` is dangling
    *ptr;
}
```

Reply:
Yes right, I'll fix it